server/src/main/java/org/elasticsearch/common/time/EpochTime.java

```diff
@@ -90,19 +124,47 @@ public TemporalAccessor resolve(
             TemporalAccessor partialTemporal,
             ResolverStyle resolverStyle
         ) {
-            long secondsAndMillis = fieldValues.remove(this);
-            long seconds = secondsAndMillis / 1_000;
-            long nanos = secondsAndMillis % 1000 * 1_000_000;
+            Long isNegative = fieldValues.remove(NEGATIVE_SIGN_FIELD);
             Long nanosOfMilli = fieldValues.remove(NANOS_OF_MILLI);
-            if (nanosOfMilli != null) {
-                nanos += nanosOfMilli;
+            long secondsAndMillis = fieldValues.remove(this);
+
+            long seconds;
+            long nanos;
+            if (isNegative != null) {
+                secondsAndMillis = -secondsAndMillis;
+                seconds = secondsAndMillis / 1_000;
+                nanos = secondsAndMillis % 1000 * 1_000_000;
+                // `secondsAndMillis < 0` implies negative timestamp; so `nanos < 0`
+                if (nanosOfMilli != null) {
+                    // aggregate fractional part of the input; subtract b/c `nanos < 0`
+                    nanos -= nanosOfMilli;
+                }
+                if (nanos != 0) {
+                    // nanos must be positive. B/c the timestamp is represented by the
+                    // (seconds, nanos) tuple, seconds moves 1s toward negative-infinity
+                    // and nanos moves 1s toward positive-infinity
+                    seconds -= 1;
+                    nanos = 1_000_000_000 + nanos;
+                }
+            } else {
+                seconds = secondsAndMillis / 1_000;
+                nanos = secondsAndMillis % 1000 * 1_000_000;
+
+                if (nanosOfMilli != null) {
+                    // aggregate fractional part of the input
+                    nanos += nanosOfMilli;
+                }
             }
+
             fieldValues.put(ChronoField.INSTANT_SECONDS, seconds);
             fieldValues.put(ChronoField.NANO_OF_SECOND, nanos);
             // if there is already a milli of second, we need to overwrite it
             if (fieldValues.containsKey(ChronoField.MILLI_OF_SECOND)) {
                 fieldValues.put(ChronoField.MILLI_OF_SECOND, nanos / 1_000_000);
             }
+            if (fieldValues.containsKey(ChronoField.MICRO_OF_SECOND)) {
```

Comment:
It is not needed. I added it defensively since we're the same logic is happening above with `MILLI_OF_SECOND`, but I'm happy to remove it since it's not directly related to the change at hand.